### ISSUE TITLE
ttnn.atanh_bw returns NaN for every element whose upstream gradient is exactly zero

### ISSUE THREAD

**Issue by ping-dev-ui** [tenstorrent/tt-metal#55290]
**Component / Area:** ttnn eltwise unary backward (composite), `atanh_bw`

**Issue Type:** Bad Outputs

## Summary

`atanh_bw` computes `grad / (1 - x^2)` and then applies

```cpp
grad_a = where(ttnn::eqz(grad), t_nan, grad_a);                       // L972
grad_a = where(logical_and(eqz(grad), eqz(input)), 0.f, grad_a);      // L973-978
```

([unary_backward.cpp#L957-L996](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L957-L996)).

The first guard was meant to encode `0 * inf = NaN` at the poles `|x| = 1`, but it has no pole condition. The second guard
rescues only `x == 0`. So every position in the open interval `0 < |x| < 1`, which is the entire finite domain of `atanh`,
returns NaN whenever the upstream gradient is exactly zero. Torch returns 0 there and NaN only at `|x| == 1`.

Zero upstream gradients are routine (attention masks, dropout, padded positions, `where` in the graph). One zero in
`grad` poisons the whole backward pass through this op.

This is a source-level analysis with a host replay; I do not have a card. The `where`/`eqz` chain is exact, so the
device result is NaN where the replay says NaN.

## Observed vs expected

| grad | x | `atanh_bw` now | torch `grad / (1 - x^2)` |
|---|---|---|---|
| 0 | 0.5 | **NaN** | 0 |
| 0 | -0.3 | **NaN** | 0 |
| 0 | 0.9 | **NaN** | 0 |
| 0 | 0.25 | **NaN** | 0 |
| 0 | 0.0 | 0 | 0 |
| 1 | 0.99 | 50.2513 | 50.2513 |
| 0 | 1.0 | NaN | NaN (0 * inf) |

## Why the tests do not catch it

The nightly test
([test_backward_atanh.py](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_atanh.py))
draws `grad` from `uniform(-100, 100)`, so an exact zero never occurs, and the PCC comparison ignores NaN lanes.

This is distinct from #54695 (the final guard rewriting overflowed `+inf` to `-inf` for negative gradients) and #53884
(dispatch count of the guards).

## Reproducer (host only, `pip install torch`)

```python
import torch
NAN, INF = float("nan"), float("inf")

def ttnn_atanh_bw(grad, x):                                   # unary_backward.cpp L957-996
    ga = grad * (1.0 / (-(x * x - 1.0)))
    ga = torch.where(grad == 0, torch.tensor(NAN), ga)                          # L972
    ga = torch.where((grad == 0) & (x == 0), torch.tensor(0.0), ga)             # L973-978
    ga = torch.where(((x == 1) | (x == -1)) & (grad != 0), torch.tensor(INF), ga)
    return torch.where((ga == INF) & (grad < 0), torch.tensor(-INF), ga)

x = torch.tensor([0.5, -0.3, 0.9, 0.0, 0.99, 0.25])
g = torch.tensor([0., 0., 0., 0., 1., 0.])
xr = x.clone().requires_grad_(True)
torch.atanh(xr).backward(g)
print(ttnn_atanh_bw(g, x).tolist())   # [nan, nan, nan, 0.0, 50.25, nan]
print(xr.grad.tolist())               # [0.0, 0.0, 0.0, 0.0, 50.25, 0.0]
```

## Proposed fix

Make the pole handling explicit and drop the unconditional guard:

```cpp
Tensor grad_a = multiply(grad, unary_chain(input, {SQUARE, SUB_UNARY_SFPU 1.0f, NEG, RECIP}));
Tensor at_pole = logical_or(eq(input, 1.0f), eq(input, -1.0f));
grad_a = where(logical_and(at_pole, eqz(grad)), t_nan, grad_a);                       // 0 * inf
grad_a = where(logical_and(at_pole, nez(grad)), multiply(sign(grad), t_inf), grad_a);  // signed inf at the poles
```

The signed-inf line also resolves #54695, since the overflow rewrite guard is no longer needed. Add a nightly case with
zeros in `grad` (for example a random 50% mask) and an explicit `|x| == 1` row.

## Acceptance criteria (proposed)

- `ttnn.atanh_bw` returns 0 wherever `grad == 0` and `|x| < 1`, matching torch autograd, for fp32 and bf16.
- NaN only where torch produces NaN (`grad == 0` at `|x| == 1`); signed infinity at `|x| == 1` for non-zero `grad`.
- Nightly test extended with a masked gradient and the pole rows.

**Comment by harmonygemini8-collab** [comment]
Requesting assignment.

🔧 Plan: The L972 guard `where(eqz(grad), NaN, grad_a)` has no pole condition — it NaNs every zero-gradient element on the whole open domain 0<|x|<1. Fix: gate the NaN on the actual poles, `where(logical_and(eqz(grad), logical_or(eq(input,1),eq(input,-1))), NaN, grad_a)`; grad==0 elsewhere then flows through `grad/(1-x²) = 0` naturally, so the L973-978 x==0 rescue becomes redundant and can be dropped. |x|=1 with grad≠0 already yields ±inf via RECIP — matches torch.

📂 Files: `ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp` (atanh_bw, ~L957-996) + unit tests with scattered zero grads over the open interval, exact poles, and x=0, golden vs torch autograd.

⏱️ ETA: 2 days. 💳 Payment: USDC (Base)

I have read and agree to the [Bounty Program Terms](https://docs.tenstorrent.com/bounty_terms.html).

**Comment by ZacLou** [comment]
I have completed root-cause analysis for this issue. The fix: gate the NaN on |x| == 1 (pole condition) instead of applying it to every zero-gradient element in 0 < |x| < 1, keeping the second guard for x == 0. Will add fp32/bf16 regression tests at boundary and zero-gradient positions. Could you assign #55290 to me?

**Comment by ZacLou** [comment]
Opened a PR that implements the fix described above: #55301.

**Comment by ZacLou** [comment]
Hi maintainers, the fix is ready in PR #55301. Could you please:`n1. Trigger the CI / approve the workflows for this first-time contribution.`n2. Confirm the bounty payout method and currency for #55290.`n`nHappy to address any review feedback promptly.

**Comment by ping-dev-ui** [comment]
Reporter here. Fix with tests is ready on my fork (https://github.com/ping-dev-ui/tt-metal/tree/fix/atanh-bw-zero-grad): both special cases are conditioned on the poles |x| == 1, the infinity takes the sign of grad directly (which also removes the overflow-rewrite guard from #54695), and nightly tests cover masked gradients and the pole rows. Holding the PR until assignment per the bounty terms. @jberkowitzTT could you assign this to me when it is reviewed?

**Comment by ZacLou** [comment]
I have read and agree to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html). PR #55301 is already open with the fix and regression tests. Could you please assign this issue to me so the bounty is tracked correctly? Thanks.

**Comment by ping-dev-ui** [comment]
I have read and agree to the Tenstorrent Bounty Program Terms & Conditions (https://docs.tenstorrent.com/bounty_terms.html).

**Comment by jberkowitzTT** [comment]
Hi @ping-dev-ui - please submit all bounty proposals to bounties@tenstorrent.com. If approved, we'll create the issue and assign it to you as a bounty

Thank you for your enthusiasm & understanding
